src/compat/db_json.cpp

```diff
@@ -0,0 +1,933 @@
+    {
+      return "JSON_ARRAY";
+    }
+  else if (document->IsObject ())
+    {
+      return "JSON_OBJECT";
+    }
+  else if (document->IsInt ())
+    {
+      return "INTEGER";
+    }
+  else if (document->IsDouble ())
+    {
+      return "DOUBLE";
+    }
+  else if (document->IsString ())
+    {
+      return "STRING";
+    }
+  else
+    {
+      /* we shouldn't get here */
+      assert (false);
+      return "UNKNOWN";
+    }
+}
+
+unsigned int
+db_json_get_length (const JSON_DOC *document)
+{
+  if (!document->IsArray () && !document->IsObject ())
+    {
+      return 1;
+    }
+  if (document->IsArray ())
+    {
+      return document->Size ();
+    }
+  if (document->IsObject ())
+    {
+      int length = 0;
+      for (JSON_VALUE::ConstMemberIterator itr = document->MemberBegin ();
+           itr != document->MemberEnd (); ++itr)
+        {
+          length++;
+        }
+
+      return length;
+    }
+
+  return 0;
+}
+
+unsigned int
+db_json_get_depth (const JSON_DOC *doc)
+{
+  return db_json_get_depth_helper (doc);
+}
+
+static unsigned int
+db_json_get_depth_helper (const JSON_VALUE *doc)
+{
+  if (doc->IsArray ())
+    {
+      unsigned int max = 0;
+      for (JSON_VALUE::ConstValueIterator itr = doc->Begin (); itr != doc->End (); ++itr)
+        {
+          unsigned int depth = db_json_get_depth_helper ((JSON_VALUE *)itr);
+          if (depth > max)
+            {
+              max = depth;
+            }
+        }
+      return max + 1;
+    }
+  else if (doc->IsObject ())
+    {
+      unsigned int max = 0;
+      for (JSON_VALUE::ConstMemberIterator itr = doc->MemberBegin (); itr != doc->MemberEnd (); ++itr)
+        {
+          unsigned int depth = db_json_get_depth_helper ((JSON_VALUE *)&itr->value);
+          if (depth > max)
+            {
+              max = depth;
+            }
+        }
+      return max + 1;
+    }
+  else
+    {
+      /* no depth */
+      return 0;
+    }
+}
+
+JSON_DOC *
+db_json_extract_document_from_path (JSON_DOC *document, const char *raw_path)
+{
+  JSON_POINTER p (raw_path);
+  JSON_VALUE *resulting_json;
+
+  if (p.IsValid () && (resulting_json = p.Get (*document)) != NULL)
+    {
+      JSON_DOC *new_doc = new JSON_DOC ();
+      new_doc->CopyFrom (*resulting_json, new_doc->GetAllocator ());
+      return new_doc;
+    }
+
+  return NULL;
+}
+
+char *
+db_json_get_raw_json_body_from_document (const JSON_DOC *doc)
+{
+  rapidjson::StringBuffer buffer;
+  rapidjson::Writer < rapidjson::StringBuffer > writer (buffer);
+  char *json_body;
+  const char *buffer_str;
+
+  buffer.Clear();
+
+  doc->Accept (writer);
+  buffer_str = buffer.GetString ();
+  json_body = (char *) db_private_alloc (NULL, strlen (buffer_str) + 1);
+  strcpy (json_body, buffer_str);
+
+  return json_body;
+}
+
+JSON_DOC *
+db_json_get_paths_for_search_func (const JSON_DOC *doc,
+                                   const char *search_str,
+                                   unsigned int one_or_all)
```

Comment:
aux is for avoiding memory corruption because rapidjson only does shallow copies (except CopyFrom). i think that if I do what you suggested (using only a doc reference value), memory corruptions will occur. I will try your suggestion and get back to you.